### ISSUE TITLE
Starting at text may not be displayed

### ISSUE THREAD

**Issue by mc12345678** [zencart/zencart#1454]
While reviewing code to answer a ZC forum question, discovered that the includes/functions/functions_lookups.php file function zen_has_attributes_values could falsely indicate that a product does not have cost modifying values assigned to it.

The function as found in ZC 1.5.5 but shown as existing since the original commit to github is:

````php function zen_has_product_attributes_values($products_id) {
    global $db;
    $attributes_query = "select sum(options_values_price) as total
                         from " . TABLE_PRODUCTS_ATTRIBUTES . "
                         where products_id = '" . (int)$products_id . "'";
    $attributes = $db->Execute($attributes_query);
    if ($attributes->fields['total'] != 0) {
      return true;
    } else {
      return false;
    }
  }
````

The problem would be visible if say there were three attributes associated with a product having a price modification of: -x, 0 and +x where x is the same "number" on both sides. Unfortunately the function doesn't also take into account other attribute related price differences that could exist independent from the "lower left corner" attribute price such as attribute quantity discounts (starting at would likely need to calculate the lowest possible price which would likely be based on ordering the largest possible quantity within the discount group), but I also think that this factor is for something else and not necessarily this function.

There is nothing in the attributes controller to force the value entered to always be positive and there probably should not be. It is possible to enter a - for the price prefix against a negative number to add the value, but regardless the question that appears to need answered by this function is: is there an attribute that can modify the price of this product when selected?

Recommend this function be rewritten as:
````php
 function zen_has_product_attributes_values($products_id) {
    global $db;
    $attributes_query = "select count(options_values_price) as total, options_values_price as price
                         from " . TABLE_PRODUCTS_ATTRIBUTES . "
                         where products_id = '" . (int)$products_id . "'";
    $attributes = $db->Execute($attributes_query);
    if ($attributes->fields['total'] == 0) {
      return false;
    } else {
      foreach ($attributes as $attribute) {
        if ($attribute['price'] != 0) {
          return true;
        } 
      }
      return false;
    }
  }
````

There are some other minor improvements that can be made to the formatting of the query and/or if the same "total" should be used or if the function should otherwise be written to support/simplify upgrading... Haven't looked at that aspect of the modification yet.

**Comment by lat9** [comment]
Since the `products_id` is a numeric value, a slight correction to the function's rewrite, removing the single-quotes around that value as shown below:

`where products_id = " . (int)$products_id;`

**Comment by mc12345678** [comment]
Yup, that was part of the other "minor improvements" not yet done. :)

**Comment by mc12345678** [comment]
Adding distinct into the query will reduce the iteration down to at most 2 tests within the loop.  At worst the first value will be 0 with a second being non-zero, otherwise if there is only one value the loop is run one time or if multiple values the first value being non-zero would exit the loop.  

Attempting to sort the data is not "guaranteed" to provide an improvement for all unless it is considered that the majority of such data would be positive and that the sort within the database server (of all potential values) as compared to the two steps on the server would provide an increase in performance.

**Comment by mc12345678** [comment]
Propose the following "improved" logic for after the query is performed and incorporating the above comments:

````php
  function zen_has_product_attributes_values($products_id) {
    global $db;
    $attributes_query = "select distinct count(options_values_price) as total, options_values_price as price
                         from " . TABLE_PRODUCTS_ATTRIBUTES . "
                         where products_id = " . (int)$products_id;

    $attributes = $db->Execute($attributes_query);

    if ($attributes->fields['total'] != 0) {
      foreach ($attributes as $attribute) {
        if ($attribute['price'] != 0) {
          return true;
        }
      }
    } 
    return false;
  }
```` 

I have generated a branch for each version of the `if` section in my fork of the repo, but I think the branches should be updated to the current version before submitting the PR(s).

**Comment by lat9** [comment]
The function is just looking for non-0.00-priced attributes, right?  Wouldn't the following query be simpler?
```
  function zen_has_product_attributes_values($products_id) {
    global $db;
    $attributes_query = "select options_values_price
                         from " . TABLE_PRODUCTS_ATTRIBUTES . "
                         where products_id = " . (int)$products_id . "
                         and options_values_price != 0
                         LIMIT 1"

    $attributes = $db->Execute($attributes_query);

    return !$attributes->EOF;
  }
```

**Comment by mc12345678** [comment]
While perhaps not specifically related to options_values_price, I have seen similar queries of "math" being performed by the database engine to provide other than desired results. Ie. a select query to provide !=0 values that when retrieved resolve to 0 but "for some reason"within the database math comparison do not resolve as expected and give a false report.  

While this tends to be more of an issue with float database field datatypes (options_values_price like other price fields is a decimal datatype), I didn't think it best to code such a situation into the base code when dealing with a non-integer number. With the newly proposed function (provided below in event of edit of the previous message), if the field were changed to a float, a string, or the database engine behaves differently with casting of an integer (0) to the field's datatype or reverse, the function may provide unsatisfactory results. This is also why I used the loose comparison to 0 instead of a strict evaluation in the php code.

Code proposed to replace previous suggestions
````php
  function zen_has_product_attributes_values($products_id) {
    global $db;
    $attributes_query = "select options_values_price
                         from " . TABLE_PRODUCTS_ATTRIBUTES . "
                         where products_id = " . (int)$products_id . "
                         and options_values_price != 0
                         LIMIT 1"

    $attributes = $db->Execute($attributes_query);

    return !$attributes->EOF;
  }
````
